Use GitHub Actions for CI instead of Travis

alouie-sfdc opened the pull request:
GitHub Actions is preferred for CI over Travis and we're moving this repo over to use GHA.

~I've attempted to do this, but am seeing that the Python 3.8 benchmark tests are consistently failing in GHA ([example](https://github.com/heroku/pghstore/actions/runs/3012880135)) and am not sure how to resolve this. They were passing in Travis and also on my local machine. I'm going to start this as a draft PR to get some feedback.~

Update: we've removed the benchmark tests from CI.

alouie-sfdc committed:
Fix some issues identified by flake8

alouie-sfdc committed:
Exclude test files from flake8 checking

We have some static analysis checks that are failing because test methods are missing docstrings. Travis CI had an `allow_failures` feature that allowed us to run static analysis checks and still report a build as passing even if static analysis fails. We're migrating to GitHub Actions and it doesn't have an analogous command, so let's explicitly exclude the two test files that are failing static analysis for now.

alouie-sfdc committed:
Move from Travis CI to GitHub Actions

alouie-sfdc committed:
Specify a pypy version

alouie-sfdc committed:
Use Python versions supported by GitHub Actions

The list of Python versions supported by GitHub Actions is at https://github.com/actions/python-versions/blob/main/versions-manifest.json

I don't think this version matters too much, because tox will create a virtualenv for the version specified in the `toxenv`.

alouie-sfdc committed:
Use pypy versions that are included in ubuntu-latest

See https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md#pypy

alouie-sfdc committed:
Update black's target version

The docs for Black 22.8.0 say that py27 is no longer a valid option:

```
  -t, --target-version [py33|py34|py35|py36|py37|py38|py39|py310|py311]
                                  Python versions that should be supported by
                                  Black's output. [default: per-file auto-
                                  detection]
```

https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html

alouie-sfdc committed:
Pin black so that we can still use it for Python 2.7

This also reverts the change from the previous commit to bring the target version back to py27.

alouie-sfdc committed:
Pin click since we pinned black

See https://stackoverflow.com/questions/71673404/importerror-cannot-import-name-unicodefun-from-click

alouie-sfdc committed:
Ignore flake8 violation caused by black

alouie-sfdc committed:
Apply black formatting changes

alouie-sfdc committed:
Use assertLess to get a better error message

This will give us the timing info instead of just "AssertionError: False is not true"

alouie-sfdc committed:
Print debug info before asserting

gsinghsfdc commented:
> They were passing in Travis and also on my local machine. 

I tried running them on my local too and it is also passing there, although I noticed that this test is benchmarking performance and even on my local the performance was pretty close b/w both for `dictlen=65` (the scenario which is failing):
```n=100000, dictlen=65, cpghstore.dumps (7.52s) is 1x faster than pghstore.dumps (7.62s)```
I'd think some failures here and there could be due to some other process but interesting that only this scenario is failing all the time on GHA.

istapletoncordascosfdc commented:
Might not be worth running benchmarks

alouie-sfdc committed:
Remove benchmark tests from CI

The benchmarks were passing in Travis CI, but we're switching to GitHub Actions and the 3.8 benchmark is consistently failing there. There are likely some external factors coming into play here, so we're going to remove the benchmarks from CI runs. They are still available to be run locally using `tox` or `tox -e <benchmark env name>`.

alouie-sfdc commented:
Thanks @istapletoncordascosfdc, I ended up removing the benchmarks from CI. Would you be able to review the PR?

kennyp commented on src/pghstore/_native.py:
```diff
@@ -177,7 +177,9 @@ def key_map(key):
-            or isinstance(value, six.binary_type)
+            or isinstance(  # noqa W503 (this is how black formats it)
+                value, six.binary_type
+            )
```
Instead of the weird formatting from `black` can't we just combine them?

```python
elif not isinstance(value, (six.string_types, six.binary_type))
```

alouie-sfdc replied:
Ah, much better! It's been updated.

kennyp reviewed (approved):
One suggestion, but everything looks good.

alouie-sfdc committed:
Use a single isinstance call

alouie-sfdc committed:
Remove unnecessary parentheses

I ran `tox -e lint` prior to my last commit, and I'm pretty sure black kept the extra parentheses. But black is still happy when I remove them.
